### PR TITLE
docs: add comprehensive Coder analysis

### PR DESCRIPTION
**Pull request by benie-joy-possi**
- Document OpenCode integration and configuration fixes
- Add VS Code Remote integration details
- Detail OAuth2 authentication setup
- Explain persistence strategies
- Add comparison and cost analysis

## 1. Summary

This PR changes:

- Add comprehensive Coder analysis documentation with architecture diagrams
- Generate PNG images from mermaid diagrams for documentation

It solves:

- Ticket #339: Understand Coder capabilities and architecture

---

## 2. Intent

The intent of this PR is:

> Document Coder's capabilities, integration points, and limitations to support the keep/discard decision for Converse platform integration.

---

### In Scope

- Coder capabilities documented: remote dev environments, workspaces, resource management
- Integration points identified: VS Code extension, OpenCode in workspace,  Known limitations and gotchas documented
- Cost analysis: infrastructure requirements, operational overhead
- Comparison with alternatives (GitPod, local dev)
- OpenCode integration findings and configuration fixes

### Out of Scope

- Implementing new Coder features or plugins
- Modifying Coder source code
- Production deployment decisions

---

## 4. Verification

I verified this change by:

- [ ] Running automated tests
- [x] Running manual tests
- [x] Checking logs
- [ ] Checking metrics
- [x] Testing error cases
- [ ] Testing permissions/security behavior
- [ ] Testing rollback or failure behavior, if relevant

Commands run:

```bash
kubectl logs coder-xxx -n coder

```

Results:

```text
 2026-06-12 15:44:19.932 [debu]  apphealth: workspace app healthy  id=054dbe4f-d80d-4796-8cac-06d99e5b8386  slug=opencode                                                             │
│ 2026-06-12 15:44:20.072 [debu]  got stats callback                                                                                                                                   │
│ 2026-06-12 15:44:20.072 [debu]  computing stats report                                                                                                                               │
│ 2026-06-12 15:44:20.072 [debu]  starting peer latency measurement for stats 
```

---

## 5. Screenshots / Evidence

Add evidence here:

image

this is a task running on the coder VS code extension

---

## 6. Risk Assessment

Risk level:

* [x] Low
* [ ] Medium
* [ ] High

Potential risks:

* Using coder requires an understanding of the right use cases

Mitigation:

* Provided this code-analysis.md   

---

## 7. AI Usage Declaration

AI was used for:

* [x] Understanding existing code
* [x] Generating code
* [ ] Refactoring
* [ ] Generating tests
* [x] Drafting documentation
* [ ] Reviewing the diff
* [ ] Not used

Human verification:

* [x] I understand every meaningful change in this PR
* [x] I checked generated code manually
* [ ] I checked generated tests manually
* [ ] I removed unsupported AI assumptions
* [x] I accept responsibility for this PR

---

## 8. Reviewer Focus

Please focus your review on:

* [x] Correctness
* [x] Architecture
* [ ] Security
* [ ] Performance
* [ ] Tests
* [x] Maintainability
* [ ] Product intent
* [ ] Edge cases

### CODE REVIEW

File: docs/coder-analysis.md [hunk 1]
```diff
@@ -0,0 +1,454 @@
+# Coder Capability & Architecture Analysis
+
+**Date:** June 12, 2026  
+**Author:** AI Governance Team
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md]: Don't you mean @benie-joy-possi ?

File: docs/coder-analysis.md [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+| OpenCode Integration | ✅ Working | OpenCode runs successfully inside Coder workspaces via the `coder-labs/opencode` module |
+| Health Check Issues | ✅ Resolved | Fixed invalid `small_model: null` configuration and model name format |
+| Authentication | ✅ Working | OAuth2 authentication via `opencode auth login` command works in workspace terminal |
+| VS Code Remote | ✅ Supported | Works seamlessly with OpenCode inside workspaces |
+| Permission System | ✅ Configurable | `ask` for interactive approval, `allow` for autonomous |
+
+### Key Value Propositions
+
+*   **Instant Onboarding:** Reduces environment setup from days to minutes.
+*   **AI Agent Orchestration:** Provides ephemeral, GPU-enabled sandboxes for AI agents (like OpenCode) to run code, test, and self-heal.
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md]: GPU enabled? Is this necessarily the case? If it isn't,  why mention it here instead of in passing?

File: docs/coder-analysis.md [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+*   **LibreChat:** AI agents running inside workspaces can access the LibreChat API via internal K8s service DNS (e.g., `http://librechat-service:3000`), provided network policies allow it.
+*   **Security:** The Coder Agent uses **outbound-only** connections, meaning no firewall ports need to be opened on the cluster for incoming workspace traffic.
+
+---
+
+## 3. VS Code Remote Integration
+
+### 3.1 How VS Code Remote Works with Coder
+
+VS Code Remote is **essential** for interactive development. When you connect via VS Code Remote, you're editing files **inside** the workspace, not on your local machine.
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md:116]: Precise here that it is a VSCode extension if that is what it is. A link to the extension somewhere would also be useful.

File: docs/coder-analysis.md [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+| **Ephemeral Test** | Pod that self-destructs after task completion. | CI/CD, security scanning, one-off AI tasks. |
+| **(Optional) VM** | Full VM via KubeVirt (if enabled). | Kernel modules, Windows, strict isolation. |
+
+### 6.2 Good Use Cases for VS Code Remote + Coder
+
+| Use Case | Why Coder Wins |
+|----------|----------------|
+| **Team Collaboration** | All developers use identical environments - no "works on my machine" issues |
+| **Secure API Key Management** | API keys stored in K8s secrets, never on developer machines; rotated centrally |
+| **Audit & Compliance** | All OpenCode sessions logged, tracked, and auditable via Coder's task reporting |
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md:538]: Task tracking huh...that's one feature we'll talk more on.

File: docs/coder-analysis.md [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+*   **Backup:** PVCs need backup strategy (Velero or native snapshots).
+
+---
+
+## 11. Conclusion & Recommendations
+
+### 11.1 Final Recommendation
+
+**Use Coder if:**
+*   You need **secure, ephemeral, GPU-rich environments** that developers cannot replicate locally.
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md:658]: GPU-rich only when we can afford it in the cloud.

File: docs/coder-analysis.md [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+*   You want **AI agents to have direct, low-latency, secure access** to code and cluster resources.
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md:659]: Like with the opencode k8s agent...interesting

File: docs/coder-analysis.md [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+
+---
+
+## Image Placeholders
+
+The following images should be created and placed in `docs/images/`:
+
+1. **coder-architecture.png** - Architecture diagram showing Control Plane / Data Plane split, user connection flow, and workspace components
+2. **vscode-remote-architecture.png** - Diagram showing how VS Code Remote connects to workspace, file access flow
+3. **opencode-health-check.png** - Health check flow between Coder Server, Coder Agent, agentapi, and opencode
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md]: Forgot to remove this.

File: docs/coder-analysis.md [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+![VS Code Remote Architecture](./images/vscode-remote-architecture.png)
+
+*The diagram above illustrates:*
+- *Your laptop runs VS Code (Editor Window)*
+- *VS Code Extension connects to Coder workspace via SSH tunnel*
+- *Files exist only in the workspace (/home/coder/project)*
+- *OpenCode runs inside workspace, sees same files as VS Code*
+- *Both edit the same remote filesystem*
+
+### 3.2 Why VS Code Remote is Essential
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md]: How about the other access mode to a coder workspace? The one where you log on to the coder ui, and then use what appears to be the VS code embedded within the coder workspace itself.

I think you should talk about this access mode too.

File: docs/coder-analysis.md [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+
+**Key Configuration Findings:**
+
+| Issue | Resolution |
+|-------|------------|
+| `small_model: null` causes crash | Remove from config or set to valid model name |
+| Model name format `provider/model` | Use correct format in `models` block (e.g., `qwen-3-5-4b-local` not `camer/qwen-3-5-4b-local`) |
+| API Key vs OAuth2 | OAuth2 preferred for production; API key works for testing |
+
+### 4.2 Authentication Setup
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md]: Documenting the OAuth flow of the AI agent in a workspace is cool, but I think we are forgetting the OAuth flow of the coder user himself? How does coder know someone has the right to create resources and workspaces and work within them?

Coder's _own_ authn/authz should be documented or at least gone over.

File: docs/coder-analysis.md [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+| **GPU Workspace** | Pod with NVIDIA GPU passthrough. | Local LLM inference, AI model training. |
+| **Multi-Container** | Pod with sidecars (e.g., DB, Redis). | Full-stack testing. |
+| **Ephemeral Test** | Pod that self-destructs after task completion. | CI/CD, security scanning, one-off AI tasks. |
+| **(Optional) VM** | Full VM via KubeVirt (if enabled). | Kernel modules, Windows, strict isolation. |
+
+### 6.2 Good Use Cases for VS Code Remote + Coder
+
+| Use Case | Why Coder Wins |
+|----------|----------------|
+| **Team Collaboration** | All developers use identical environments - no "works on my machine" issues |
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md:535]: I do get the sentiment, but there are a few reasons this doesn't hold up:
- the difference in environment between developers running and testing their code locally is small, and its relevance is reduced even further by containerization tools like docker(and the fact that most apps today are containerized).
- let's not forget also that the size of a coder workspace is much more limited compared to a local machine, and that the resources coder workspaces consume cost money(on the cloud) as opposed to, say, apps and IDEs that run locally. Imagine a coder workspace running VSCode 8hrs straight in it - that's a lot of compute.

The only real advantage that I see in using a coder workspace for development is the network condition in the cloud vs locally.

**Review by Koufan-De-King — CHANGES_REQUESTED** [review]

File: docs/coder-analysis.md (hunk 5 above)
Comment (benie-joy-possi) [reply docs/coder-analysis.md:658]: Yeah true!

File: docs/coder-analysis.md (hunk 10 above)
Comment (benie-joy-possi) [reply docs/coder-analysis.md:535]: Fair and valid points there! In addition to the network condition we have the AI Agent Scale(parallelism). But these  points you made are also things I will keep in mind  when working  on ticket #353.

File: docs/coder-analysis.md (hunk 3 above)
Comment (benie-joy-possi) [reply docs/coder-analysis.md:116]: I will elaborate on that

File: docs/coder-analysis.md [hunk 11]
```diff
@@ -0,0 +1,454 @@
...
+**Author:** AI Governance Team  
+**Status:** Draft for Review  
+**Ticket:** #339 (Spike: Understand Coder capabilities and architecture)  
+**Context:** Evaluation of Coder for integration into the **Converse** AI platform (LibreChat, Envoy Gateway ecosystem).
+
+---
+
+## 1. Executive Summary
+
+**Recommendation:** **Proceed with Production Deployment.**
```
[the remark this answers is not in the document]
Comment (benie-joy-possi) [reply docs/coder-analysis.md]: Investigations are still being made to later produce a verdict on whether it will be reintroduce in the our architecture from there a new ADR will be added

File: docs/coder-analysis.md (hunk 8 above)
Comment (benie-joy-possi) [reply docs/coder-analysis.md]: It is just our vs code desktop  but then an ssh connection is created where we can acess a coder worspace

File: docs/coder-analysis.md [hunk 12]
```diff
@@ -0,0 +1,696 @@
...
+    A[User] --> B[VS Code Desktop]
+    A --> C[VS Code Web<br/>code-server]
+    A --> D[SSH Terminal]
+    
+    B --> E[SSH Tunnel via Coder Agent]
+    C --> F[HTTP/WebSocket via Coder Proxy]
+    D --> E
+    
+    E --> G[Workspace Pod]
+    F --> G
```
Comment (Koufan-De-King) [inline docs/coder-analysis.md:247]: There is a problem here
Image

**Review by Koufan-De-King — APPROVED** [review]

File: docs/coder-analysis.md (hunk 12 above)
Comment (benie-joy-possi) [reply docs/coder-analysis.md:247]: but I can see it works when i click on `Display rich diff`  and on mermaid.live too

Comment (benie-joy-possi) [reply docs/coder-analysis.md:247]: image

Comment (Koufan-De-King) [reply docs/coder-analysis.md:247]: Probably network issues on my side :+1:
